apps/dashboard/src/context/environment/environment-provider.tsx

```diff
@@ -35,18 +32,51 @@ function selectEnvironment(environments?: IEnvironment[], selectedEnvironmentId?
 
 export function EnvironmentProvider({ children }: { children: React.ReactNode }) {
   const { currentOrganization } = useAuth();
+  const navigate = useNavigate();
+  const { pathname } = useLocation();
+  const { environmentId: paramsEnvironmentId } = useParams<{ environmentId?: string }>();
   const [currentEnvironment, setCurrentEnvironment] = useState<IEnvironment | null>(null);
 
+  const switchEnvironmentInternal = useCallback(
+    (allEnvironments: IEnvironment[], environmentId?: string | null) => {
+      const selectedEnvironment = selectEnvironment(allEnvironments, environmentId);
+      setCurrentEnvironment(selectedEnvironment);
+      const newEnvironmentId = selectedEnvironment._id;
+      const isNewEnvironmentDifferent = paramsEnvironmentId !== selectedEnvironment._id;
+
+      if (pathname === ROUTES.ROOT || pathname === ROUTES.ENV || pathname === `${ROUTES.ENV}/`) {
+        navigate(buildRoute(ROUTES.WORKFLOWS, { environmentId: newEnvironmentId }));
+      } else if (pathname.includes(ROUTES.ENV) && isNewEnvironmentDifferent) {
+        const newPath = pathname.replace(/\/env\/[^/]+(\/|$)/, `${ROUTES.ENV}/${newEnvironmentId}$1`);
+        navigate(newPath);
```

Comment:
if URL includes `/env` and env id in the URL is different from the "selected" (from all envs), then update the URL

Reply:
I am not sure if this will always work. If the URL contains more ids, for example, the workflow slug or the step slug, switching across environments will return a 404 errors as the workflow or step id won't match the current environment. 

My suggestion is to always go to `/env/:environmentId/workflows` after switching environment.

Reply:
@SokratisVidros it works with all these cases:

![Screenshot 2024-10-07 at 18 12 12](https://github.com/user-attachments/assets/753d6106-23d1-4301-bdd3-5fec1c713590)

---

IMO it's important to keep the user on the exact same page where he was, otherwise, it will be a frustrating experience, especially if the user is somewhere two levels deep. And this logic is intended to do that.

If we want to cover the case when the workflow or step doesn't exist, I think that we should have the redirect logic in the corresponding places where we make requests, and when it fails redirect one level down, until we go to the root route. 

For example, if the user received from someone the wrong URL (or switched the env) `/env/:environmentId/workflows/:workflowId/steps/:stepId`
We can have the logic that:
- load `:stepId` when it fails redirect to `/env/:environmentId/workflows/:workflowId`
- load `:workflowId` when it fails redirect to `/env/:environmentId/workflows`
- `:environmentId` doesn't exist - switch to develop

This is like a waterfall effect, but IMO, it will provide better UX to cover similar cases as the application "will try to do best" to render the closest page to the context.

Reply:
The `workflowId` and `stepId` will always be different between environments. Let's take it offline tomorrow.